File: scripts/evaluate_practice_alignment.mjs
import { createHash } from "node:crypto";
import { readFileSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

import { practiceComparisonAlignment } from "../cloudflare/worker.mjs";

function parseArgs(argv) {
  const fixtures = [];
  const excluded = new Set();
  let output = "";
  let summaryOnly = false;
  for (let index = 0; index < argv.length; index += 1) {
    const value = argv[index];
    if (value === "--exclude") {
      excluded.add(String(argv[index += 1] || ""));
    } else if (value === "--output") {
      output = String(argv[index += 1] || "");
    } else if (value === "--summary-only") {
      summaryOnly = true;
    } else {
      fixtures.push(value);
    }
  }
  if (!fixtures.length) {
    throw new Error("at least one fixture path is required");
  }
  return { fixtures, excluded, output, summaryOnly };
}

function sameTimestamp(actual, expected) {
  if (actual === null || actual === undefined || expected === null || expected === undefined) {
    return (actual === null || actual === undefined) && (expected === null || expected === undefined);
  }
  return Math.abs(Number(actual) - Number(expected)) < 1e-6;
}

function compareCase(fixture) {
  const actual = practiceComparisonAlignment({
    targetText: fixture.target_text,
    recognizedText: fixture.recognized_text,
    targetLanguage: fixture.target_language,
    asrTimestamps: fixture.asr_timestamps,
  });
  const expected = fixture.expected;
  const mismatches = [];
  for (const key of ["available", "complete"]) {
    if (actual[key] !== expected[key]) {
      mismatches.push({ field: key, expected: expected[key], actual: actual[key] });
    }
  }
  if (actual.ranges.length !== expected.ranges.length) {
    mismatches.push({ field: "range_count", expected: expected.ranges.length, actual: actual.ranges.length });
  }
  for (let index = 0; index < Math.min(actual.ranges.length, expected.ranges.length); index += 1) {
    const actualRange = actual.ranges[index];
    const expectedRange = expected.ranges[index];
    for (const key of ["index", "source", "available", "matched_text"]) {
      if (actualRange[key] !== expectedRange[key]) {
        mismatches.push({ field: `ranges[${index}].${key}`, expected: expectedRange[key], actual: actualRange[key] });
      }
    }
    for (const key of ["audio_start", "audio_end"]) {
      if (!sameTimestamp(actualRange[key], expectedRange[key])) {
        mismatches.push({ field: `ranges[${index}].${key}`, expected: expectedRange[key], actual: actualRange[key] });
      }
    }
  }
  return {
    name: fixture.name,
    target_language: fixture.target_language,
    category: fixture.category || "",
    passed: mismatches.length === 0,
    mismatches,
    actual,
  };
}

const args = parseArgs(process.argv.slice(2));
const fixtureHashes = {};
const cases = args.fixtures.flatMap((fixture) => {
  const source = readFileSync(fixture);
  fixtureHashes[fixture] = createHash("sha256").update(source).digest("hex");
  const loaded = JSON.parse(source.toString("utf8"));
  if (!Array.isArray(loaded)) {
    throw new Error(`fixture must contain a JSON array: ${fixture}`);
  }
  return loaded;
});
const results = cases.map(compareCase);
const evaluated = results.filter((result) => !args.excluded.has(result.name));
const payload = {
  runtime: "cloudflare-worker",
  fixtures: fixtureHashes,
  total: results.length,
  evaluated: evaluated.length,
  excluded: results.filter((result) => args.excluded.has(result.name)).map((result) => result.name),
  passed: evaluated.filter((result) => result.passed).length,
  failed: evaluated.filter((result) => !result.passed).length,
  results,
};
if (args.output) {
  mkdirSync(dirname(args.output), { recursive: true });
  writeFileSync(args.output, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
}
const summary = Object.fromEntries(["runtime", "total", "evaluated", "passed", "failed", "excluded"].map((key) => [key, payload[key]]));
process.stdout.write(`${JSON.stringify(args.summaryOnly ? summary : payload)}\n`);

File: scripts/evaluate_practice_alignment.py
from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path
from typing import Any

from mo_speech.practice import practice_comparison_alignment


def _same_timestamp(actual: object, expected: object) -> bool:
    if actual is None or expected is None:
        return actual is None and expected is None
    return abs(float(actual) - float(expected)) < 1e-6


def _compare_case(case: dict[str, Any]) -> dict[str, Any]:
    actual = practice_comparison_alignment(
        target_text=str(case["target_text"]),
        recognized_text=str(case["recognized_text"]),
        target_language=str(case["target_language"]),
        asr_timestamps=case["asr_timestamps"],
    )
    expected = case["expected"]
    mismatches: list[dict[str, object]] = []
    for key in ("available", "complete"):
        if actual.get(key) != expected.get(key):
            mismatches.append({"field": key, "expected": expected.get(key), "actual": actual.get(key)})
    actual_ranges = list(actual.get("ranges") or [])
    expected_ranges = list(expected.get("ranges") or [])
    if len(actual_ranges) != len(expected_ranges):
        mismatches.append({"field": "range_count", "expected": len(expected_ranges), "actual": len(actual_ranges)})
    for index, (actual_range, expected_range) in enumerate(zip(actual_ranges, expected_ranges)):
        for key in ("index", "source", "available", "matched_text"):
            if actual_range.get(key) != expected_range.get(key):
                mismatches.append(
                    {
                        "field": f"ranges[{index}].{key}",
                        "expected": expected_range.get(key),
                        "actual": actual_range.get(key),
                    }
                )
        for key in ("audio_start", "audio_end"):
            if not _same_timestamp(actual_range.get(key), expected_range.get(key)):
                mismatches.append(
                    {
                        "field": f"ranges[{index}].{key}",
                        "expected": expected_range.get(key),
                        "actual": actual_range.get(key),
                    }
                )
    return {
        "name": case["name"],
        "target_language": case["target_language"],
        "category": case.get("category", ""),
        "passed": not mismatches,
        "mismatches": mismatches,
        "actual": actual,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Evaluate SpeakLoop alignment fixtures with the Python runtime.")
    parser.add_argument("fixtures", nargs="+", type=Path)
    parser.add_argument("--exclude", action="append", default=[], help="case name to report but exclude from pass/fail totals")
    parser.add_argument("--output", type=Path)
    parser.add_argument("--summary-only", action="store_true")
    args = parser.parse_args()

    cases: list[dict[str, Any]] = []
    fixture_hashes: dict[str, str] = {}
    for fixture in args.fixtures:
        payload = fixture.read_bytes()
        fixture_hashes[str(fixture)] = hashlib.sha256(payload).hexdigest()
        loaded = json.loads(payload)
        if not isinstance(loaded, list):
            raise ValueError(f"fixture must contain a JSON array: {fixture}")
        cases.extend(loaded)

    excluded_names = set(args.exclude)
    results = [_compare_case(case) for case in cases]
    evaluated = [result for result in results if result["name"] not in excluded_names]
    payload = {
        "runtime": "python",
        "fixtures": fixture_hashes,
        "total": len(results),
        "evaluated": len(evaluated),
        "excluded": [result["name"] for result in results if result["name"] in excluded_names],
        "passed": sum(bool(result["passed"]) for result in evaluated),
        "failed": sum(not bool(result["passed"]) for result in evaluated),
        "results": results,
    }
    if args.output:
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    summary = {key: payload[key] for key in ("runtime", "total", "evaluated", "passed", "failed", "excluded")}
    print(json.dumps(summary if args.summary_only else payload, ensure_ascii=False))


if __name__ == "__main__":
    main()
